### PR TITLE
Pubsub unified function

### PR DESCRIPTION
**Pull request by archishou**
This accomplishes the same thing as PR #615 but I had to rename the branch so this is the new PR

### CODE REVIEW

**Comment by archishou** [comment]
For the Subscription side, there is already a SubscriptionOptions struct being passed into the subscribeImpl function. Should I add new fields to the SubscriptionOptions so that this struct is the only thing being passed into subscribeImpl? Or is there a different way you think could be better?

**Comment by deitch** [comment]
> Should I add new fields to the SubscriptionOptions so that this struct is the only thing being passed into subscribeImpl? 

Yes, I think your suggestion is good. You can roll `agentName`, `agentScope`, `persistent`, `topicType`, and `ctx` into the `SubscriptionOptions` for a nice clean interface.

I just am torn on the `activate` one. On the one hand, I like the idea of a single parameter to pass. On the other hand, all of the other parameters are _characteristics_ of the queue - what kind is it, topic, scope, persistence, handlers, etc., while the `activate` one is imperative: "take this action".

You know what? Do with `activate` whatever you want. I just convinced myself with the above paragraph that we should not even have that option, but should do it in 2 steps.

Once your PR is complete and done, we can do a PR moving it outside.

This is nice and clean, I like it. :-)

**Comment by archishou** [comment]
Thank you! Last thing, though, for both subscription and publication is there a reason subscribeImpl and publishImpl aren't public? If we are just passing in this one struct, the wrapper function that just parses the fields from the struct feels like unnecessary as the subscribeImpl and publishImpl could just handle that on their own. Basically, I'm wondering if there is any reason we should keep the new Publication and Subscription functions since all these functions would be doing is calling subscribeImpl and publishImpl if it determines that values for the fields make sense, something I feel could be handled by subscribeImpl and publishImpl

**Comment by deitch** [comment]
I cannot think of a good reason why not. Just:

* make sure it isn't called anywhere else (it shouldn't be, since the big restructure I did back in late December, but worth checking)
* add your checks for missing components that you put in `Publication()`
* rename `subscribeImpl` to `Subscription()` and `publishImpl` to `Publication()`; your naming is _much_ better to expose than those old ones :-)

**Comment by archishou** [comment]
Ok, I added it. I added the same checks to subscription that I did to publication. I think those are sufficient since that's essentially what the old Subscribe, SubscribePersitent, and SubscribeScope used to do. Please let me know if I am missing anything, thank you

**Comment by deitch** [comment]
This is really nice, much better. I would not merge it without running a full "boot the system and download test", but quite good. One more requested change, just a nit. Will comment.

**Comment by archishou** [comment]
I'm sorry if this is obvious, but how do I do a full boot-and-run test?

**Comment by deitch** [comment]
Build, run, boot, see everything comes up, download and run a container and a vm. @eriknordmark has some integration tests for it.

**Comment by kalyan-nidumolu** [comment]
Hi Archishmaan, Let me take it up to run those tests. You don't have access to the devices needed to run the test.

**Comment by kalyan-nidumolu** [comment]
Ok.. A nit.. I am thinking we should change:
   1) pubsub.Publish to pubsub.NewPublisher()

Publish is confusing - they refer to something being published - rather than a publisher being created.

**Comment by deitch** [comment]
> pubsub.Publish to pubsub.NewPublisher

He did; he changed `pubsub.Publish`/`pubsub.PublishScope`/`pubsub.PublishPersistent` to just `pubsub.Publication`.

If you mean that it should be `pubsub.Publisher` instead of `pubsub.Publication` [here])(https://github.com/lf-edge/eve/pull/616/files#diff-4e6d1d96b37fb7294eca579a2cb30de8R146) and `pubsub.Subscriber` instead of `pubsub.Subscription` [here](https://github.com/lf-edge/eve/pull/616/files#diff-4e6d1d96b37fb7294eca579a2cb30de8R82), then I don't really care much of which name we use.

**Comment by archishou** [comment]
The current names are Publication and Subscription. Should I be changing the names to 
NewPublication and NewSubscription or Publisher and Subscriber

**Comment by deitch** [comment]
The renaming is fine. I actually liked the previous one better `func (p* pubsub) Publication()` instead of `func (p *pubsub) NewPublication()`, but that is 100% personal preference, so it just doesn't matter.

However, with the current code, the code itself is correct, but the comments are off now. I will point those out inline.

File: pkg/pillar/pubsub/doc.go [hunk 1]
```diff
@@ -50,13 +50,13 @@
 //
 // The driver is responsible for implementing the underlying mechanics of
 // publishing and subscribing. While `pubsub.PubSub` and its components -
-// `Publication` and `Subscription` - handle the in-memory and diff aspects,
+// `NewPublication` and `NewSubscription` - handle the in-memory and diff aspects,
```
Comment (deitch) [inline pkg/pillar/pubsub/doc.go]: No, this is wrong. `Publication` and `Subscription` handle the in-memory... 

`NewPublication` and `NewSubscription` are just the function calls that _create_ the `Publication` and `Subscription`, which handle the in-memory and...

File: pkg/pillar/pubsub/doc.go [hunk 2]
```diff
@@ -50,13 +50,13 @@
...
 // the driver itself is responsible for communicating between the publisher
 // and subscriber, and performing any persistence.
 //
 // The driver is expected to implement the `Driver` interface, which primarily
 // involves being able to return the `DriverPublisher` and `DriverSubscriber`.
-// These in turn are used by `Publication` and `Subscription` to publish and
+// These in turn are used by `NewPublication` and `NewSubscription` to publish and
```
Comment (deitch) [inline pkg/pillar/pubsub/doc.go]: Again, revert this change. It is the `Publication` and `Subscription` that do the publishing and subscribing; `NewSub` and `NewPub` are just the func calls to create them.

File: pkg/pillar/pubsub/driver.go [hunk 3]
```diff
@@ -52,7 +52,7 @@ type DriverPublisher interface {
 	Restart(restarted bool) error
 }
 
-// Restarted interface that lets you determine if a Publication has been restarted
+// Restarted interface that lets you determine if a NewPublication has been restarted
```
Comment (deitch) [inline pkg/pillar/pubsub/driver.go]: Revert. It is the `Publication` that has been restarted; `NewPublication` is just the func call to create the `Publication`

File: pkg/pillar/pubsub/legacy/legacy.go [hunk 4]
```diff
@@ -10,51 +10,102 @@ var (
 	defaultPubsub = pubsub.New(&socketdriver.SocketDriver{})
 )
 
-// Publish create a `Publication` for the given agent name and topic type.
+// Publish create a `NewPublication` for the given agent name and topic type.
```
Comment (deitch) [inline pkg/pillar/pubsub/legacy/legacy.go]: Same as before. Revert comment.

File: pkg/pillar/pubsub/legacy/legacy.go [hunk 5]
```diff
@@ -10,51 +10,102 @@ var (
...
 	log.Infof("legacy.Publish agentName(%s)", agentName)
-	return defaultPubsub.Publish(agentName, topicType)
+	return defaultPubsub.NewPublication(&pubsub.PublicationOptions{
+		AgentName: agentName,
+		TopicType: topicType,
+	})
 }
 
-// PublishPersistent create a `Publication` for the given agent name and topic
+// PublishPersistent create a `NewPublication` for the given agent name and topic
```
Comment (deitch) [inline pkg/pillar/pubsub/legacy/legacy.go]: Same as before; revert comment.

File: pkg/pillar/pubsub/legacy/legacy.go [hunk 6]
```diff
@@ -10,51 +10,102 @@ var (
...
 // type, but with persistence of the messages across reboots.
 func PublishPersistent(agentName string, topicType interface{}) (pubsub.Publication, error) {
 	log.Infof("legacy.PublishPersistent agentName(%s)", agentName)
-	return defaultPubsub.PublishPersistent(agentName, topicType)
+	return defaultPubsub.NewPublication(&pubsub.PublicationOptions{
```
Comment (deitch) [inline pkg/pillar/pubsub/legacy/legacy.go]: This is correct.

File: pkg/pillar/pubsub/legacy/legacy.go [hunk 7]
```diff
@@ -10,51 +10,102 @@ var (
...
+		AgentName:  agentName,
+		TopicType:  topicType,
+		Persistent: true,
+	})
 }
 
-// PublishScope create a `Publication` for the given agent name and topic,
+// PublishScope create a `NewPublication` for the given agent name and topic,
```
Comment (deitch) [inline pkg/pillar/pubsub/legacy/legacy.go]: Same as before, revert the comment. It is creating a `Publication` (the thing it is returning), not a `NewPublication` (the func call)

File: pkg/pillar/pubsub/legacy/legacy.go [hunk 8]
```diff
@@ -10,51 +10,102 @@ var (
...
+		TopicType:  topicType,
+		AgentScope: agentScope,
+	})
 }
 
 // Subscribe create a subscription for the given agent name and topic
 // optionally activating immediately. If `activate` is set to `false`
 // (the default), then the subscription will not begin to send messages
-// on the channel or process them until `Subscription.Start()` is called.
+// on the channel or process them until `NewSubscription.Start()` is called.
```
Comment (deitch) [inline pkg/pillar/pubsub/legacy/legacy.go]: Same as before; revert comment

File: pkg/pillar/pubsub/legacy/legacy.go [hunk 9]
```diff
@@ -10,51 +10,102 @@ var (
...
+		Persistent:     false,
+	})
 }
 
 // SubscribeScope create a subscription for the given agent name and topic,
 // limited to a given scope,
 // optionally activating immediately. If `activate` is set to `false`
 // (the default), then the subscription will not begin to send messages
-// on the channel or process them until `Subscription.Start()` is called.
+// on the channel or process them until `NewSubscription.Start()` is called.
```
Comment (deitch) [inline pkg/pillar/pubsub/legacy/legacy.go]: Same as before; revert comment

File: pkg/pillar/pubsub/legacy/legacy.go [hunk 10]
```diff
@@ -10,51 +10,102 @@ var (
...
 func SubscribeScope(agentName string, agentScope string, topicType interface{},
 	activate bool, ctx interface{}, options *pubsub.SubscriptionOptions) (pubsub.Subscription, error) {
-	return defaultPubsub.SubscribeScope(agentName, agentScope, topicType, activate, ctx, options)
+	return defaultPubsub.NewSubscription(&pubsub.SubscriptionOptions{
```
Comment (deitch) [inline pkg/pillar/pubsub/legacy/legacy.go]: This is good.

File: pkg/pillar/pubsub/legacy/legacy.go [hunk 11]
```diff
@@ -10,51 +10,102 @@ var (
...
+		Persistent:     false,
+	})
 }
 
 // SubscribePersistent create a subscription for the given agent name and topic,
 // persistent,
 // optionally activating immediately. If `activate` is set to `false`
 // (the default), then the subscription will not begin to send messages
-// on the channel or process them until `Subscription.Start()` is called.
+// on the channel or process them until `NewSubscription.Start()` is called.
```
Comment (deitch) [inline pkg/pillar/pubsub/legacy/legacy.go]: Same as before; revert comment.

File: pkg/pillar/pubsub/legacy/legacy.go [hunk 12]
```diff
@@ -10,51 +10,102 @@ var (
...
 func SubscribePersistent(agentName string, topicType interface{}, activate bool,
 	ctx interface{}, options *pubsub.SubscriptionOptions) (pubsub.Subscription, error) {
-	return defaultPubsub.SubscribePersistent(agentName, topicType, activate, ctx, options)
+	return defaultPubsub.NewSubscription(&pubsub.SubscriptionOptions{
```
Comment (deitch) [inline pkg/pillar/pubsub/legacy/legacy.go]: This is good.

File: pkg/pillar/pubsub/publish.go [hunk 13]
```diff
@@ -22,7 +22,7 @@ type LocalCollection map[string][]byte
 // Notify simple struct to pass notification messages
 type Notify struct{}
 
-// Publication to publish to an individual topic
+// NewPublication to publish to an individual topic
```
Comment (deitch) [inline pkg/pillar/pubsub/publish.go]: Same as before, revert comment

File: pkg/pillar/pubsub/publish.go [hunk 14]
```diff
@@ -35,8 +35,8 @@ type Notify struct{}
 //  foo := p1.Get(key)
 //  fooAll := p1.GetAll()
 
-// PublicationImpl - Publication Implementation. The main structure that implements
-//  Publication interface.
+// PublicationImpl - NewPublication Implementation. The main structure that implements
+//  NewPublication interface.
```
Comment (deitch) [inline pkg/pillar/pubsub/publish.go]: Same as before, revert comment. This implements `Publication interface`, the interface, not `NewPublication`, the func call.

File: pkg/pillar/pubsub/pubsub.go [hunk 15]
```diff
@@ -7,13 +7,14 @@
...
 import (
+	"fmt"
 	"reflect"
 	"time"
 
 	log "github.com/sirupsen/logrus"
 )
 
-// SubscriptionOptions options to pass when creating a Subscription
+// SubscriptionOptions options to pass when creating a NewSubscription
```
Comment (deitch) [inline pkg/pillar/pubsub/pubsub.go]: Same as before; revert comment

File: pkg/pillar/pubsub/pubsub.go [hunk 16]
```diff
@@ -53,7 +60,7 @@ type keyMap struct {
 }
 
 // PubSub is a system for publishing and subscribing to messages
-// it manages the creation of Publication and Subscription, which handle the actual
+// it manages the creation of NewPublication and NewSubscription, which handle the actual
```
Comment (deitch) [inline pkg/pillar/pubsub/pubsub.go]: Same as before, revert comment. We are creating a `Publication` or a `Subscription`, the things; `NewPublication` and `NewSubscription` are just func calls that create, respectively, `Publication` and `Subscription`.

File: pkg/pillar/pubsub/pubsub.go [hunk 17]
```diff
@@ -163,29 +133,48 @@ func (p *PubSub) subscribeImpl(agentName string, agentScope string, topicImpl in
...
-	topic := TypeToName(topicType)
+// PublicationOptions defines all the possible options a new publication may have
+type PublicationOptions struct {
+	AgentName  string
+	AgentScope string
+	TopicType  interface{}
+	Persistent bool
+}
+
+// NewPublication creates a new publication with given options
```
Comment (deitch) [inline pkg/pillar/pubsub/pubsub.go]: Small nit, but should capitalize: `creates a new Publication` (that way it is clear that we are referring to the interface returned)

File: pkg/pillar/pubsub/pubsubintf.go [hunk 18]
```diff
@@ -5,7 +5,7 @@
 
 package pubsub
 
-// Publication - Interface to be implemented by a Publication
+// Publication - Interface to be implemented by a NewPublication
```
Comment (deitch) [inline pkg/pillar/pubsub/pubsubintf.go]: Same as before; revert comment. A `Publication` implementation implements the interface; `NewPublication` is just the func call.

File: pkg/pillar/pubsub/pubsubintf.go [hunk 19]
```diff
@@ -23,7 +23,7 @@ type Publication interface {
 	Iterate(function fn)
 }
 
-// Subscription - Interface to be implemented by a Subscription
+// Subscription - Interface to be implemented by a NewSubscription
```
Comment (deitch) [inline pkg/pillar/pubsub/pubsubintf.go]: Same as before; revert the comment

File: pkg/pillar/pubsub/pubsubintf.go [hunk 20]
```diff
@@ -33,9 +33,9 @@ type Subscription interface {
 	Iterate(function fn)
 	// Restarted report if this subscription has been marked as restarted
 	Restarted() bool
-	// ProcessChange - Invoked on the string msg from Subscription Channel
+	// ProcessChange - Invoked on the string msg from NewSubscription Channel
```
Comment (deitch) [inline pkg/pillar/pubsub/pubsubintf.go]: Same as before; revert the comment

File: pkg/pillar/pubsub/pubsubintf.go [hunk 21]
```diff
@@ -33,9 +33,9 @@ type Subscription interface {
...
 	ProcessChange(change Change)
-	// MsgChan - Message Channel for Subscription
+	// MsgChan - Message Channel for NewSubscription
```
Comment (deitch) [inline pkg/pillar/pubsub/pubsubintf.go]: Same as before; revert the comment

File: pkg/pillar/pubsub/subscribe.go [hunk 22]
```diff
@@ -36,7 +36,7 @@ type SubscriptionImpl struct {
 	defaultName  string
 }
 
-// MsgChan return the Message Channel for the Subscription.
+// MsgChan return the Message Channel for the NewSubscription.
```
Comment (deitch) [inline pkg/pillar/pubsub/subscribe.go]: Same as before; revert the comment

File: pkg/pillar/pubsub/subscribe.go [hunk 23]
```diff
@@ -68,7 +68,7 @@ func (sub *SubscriptionImpl) ProcessChange(change Change) {
 	CheckMaxTimeTopic(sub.agentName, sub.topic, start, sub.MaxProcessTimeWarn, sub.MaxProcessTimeError)
 }
 
-// Get - Get object with specified Key from this Subscription.
+// Get - Get object with specified Key from this NewSubscription.
```
Comment (deitch) [inline pkg/pillar/pubsub/subscribe.go]: Same as before, revert the comment

**Comment by kalyan-nidumolu** [comment]
Hi Archishmaan - There are currently 5 commits in this PR.. Can you squash them into 1 and put appropriate commit comments?

**Comment by kalyan-nidumolu** [comment]
FYI.. Here is the Error I am seeing - agents failing to Publish or Subscribe - Debugging it:

Reboot from agent ledmanager at 2020-02-12T07:53:46.865049795Z: fatal: agent ledmanager: cannot create a subcription with a nil agentName
Reboot from agent ledmanager at 2020-02-12T07:53:46.877394329Z: fatal: agent ledmanager exit
Reboot from agent nodeagent at 2020-02-12T07:53:49.281145844Z: fatal: agent nodeagent: cannot create a publication with a nil agentName
Reboot from agent nodeagent at 2020-02-12T07:53:49.2871144Z: fatal: agent nodeagent exit
Reboot from agent nim at 2020-02-12T07:54:49.779232627Z: fatal: agent nim: cannot create a publication with a nil agentName
Reboot from agent nim at 2020-02-12T07:54:49.784336916Z: fatal: agent nim exit
Reboot from agent diag at 2020-02-12T07:54:52.910585448Z: fatal: agent diag: cannot create a publication with a nil agentName
Reboot from agent diag at 2020-02-12T07:54:52.918774258Z: fatal: agent diag exit

**Comment by archishou** [comment]
I will remove the check.

File: pkg/pillar/pubsub/pubsub.go [hunk 24]
```diff
@@ -69,68 +76,31 @@ func New(driver Driver) *PubSub {
...
-}
-
 // methods unique to this implementation
 
-func (p *PubSub) subscribeImpl(agentName string, agentScope string, topicImpl interface{},
-	activate bool, ctx interface{}, persistent bool, options *SubscriptionOptions) (Subscription, error) {
+// NewSubscription creates a new Subscription with given options
+func (p *PubSub) NewSubscription(options *SubscriptionOptions) (Subscription, error) {
+
+	if options.AgentScope != "" && options.Persistent == true {
```
Comment (eriknordmark) [inline pkg/pillar/pubsub/pubsub.go]: While we don't use this, I see no reason to ban it.

File: pkg/pillar/pubsub/pubsub.go [hunk 25]
```diff
@@ -69,68 +76,31 @@ func New(driver Driver) *PubSub {
...
+		return nil, fmt.Errorf("cannot create a persitent subcription with a scope agentName %s", options.AgentName)
+	}
+	if options.AgentName == "" {
```
Comment (eriknordmark) [inline pkg/pillar/pubsub/pubsub.go]: This is the unneeded check causing failures I guess.

File: pkg/pillar/pubsub/pubsub.go [hunk 26]
```diff
@@ -163,29 +133,47 @@ func (p *PubSub) subscribeImpl(agentName string, agentScope string, topicImpl in
...
+type PublicationOptions struct {
+	AgentName  string
+	AgentScope string
+	TopicType  interface{}
+	Persistent bool
+}
+
+// NewPublication creates a new Publication with given options
+func (p *PubSub) NewPublication(options *PublicationOptions) (Publication, error) {
+	if options.AgentScope != "" && options.Persistent == true {
```
Comment (eriknordmark) [inline pkg/pillar/pubsub/pubsub.go]: No reason to prohibit this.

**Comment by kalyan-nidumolu** [comment]
Fixed a few issues.. Invalid checks, Some Pointer issues..

1) I made NewSubscription and NewPublication take Options as opposed to OptionsPtr
2) Took care of Erik's comments about un-needed checks

**Comment by eriknordmark** [comment]
> Fixed a few issues.. Invalid checks, Some Pointer issues..
> 
> 1. I made NewSubscription and NewPublication take Options as opposed to OptionsPtr
> 2. Took care of Erik's comments about un-needed checks

@kalyan-nidumolu You are still listed as "changes requested" so I'm not pulling this quite yet.
Let me know when testing has passed and we can pull.

**Comment by kalyan-nidumolu** [comment]
Ok.. All issues fixed.. Test one app also passed..This one is ready to be merged..
